src/test/spec/unified_runner/test_file.rs

```diff
@@ -0,0 +1,369 @@
+                return false;
+            }
+        }
+        if let Some(ref max_version) = self.max_server_version {
+            let req = VersionReq::parse(&format!("<= {}", &max_version)).unwrap();
+            if !req.matches(&client.server_version) {
+                return false;
+            }
+        }
+        if let Some(ref topologies) = self.topologies {
+            if !topologies.contains(&client.topology()) {
+                return false;
+            }
+        }
+        true
+    }
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub enum Entity {
+    Client(Client),
+    Database(Database),
+    Collection(Collection),
+    Session(Session),
+    Bucket(Bucket),
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Client {
+    pub id: String,
+    #[serde(
+        default = "default_uri",
+        deserialize_with = "deserialize_uri_options_to_uri_string",
+        rename = "uriOptions"
+    )]
+    pub uri: String,
+    pub use_multiple_mongoses: Option<bool>,
+    pub observe_events: Option<Vec<String>>,
+    pub ignore_command_monitoring_events: Option<Vec<String>>,
+    #[serde(default)]
+    #[serde(deserialize_with = "deserialize_server_api")]
+    pub server_api: Option<ServerApi>,
+}
+
+fn default_uri() -> String {
+    DEFAULT_URI.clone()
+}
+
+fn deserialize_uri_options_to_uri_string<'de, D>(
+    deserializer: D,
+) -> std::result::Result<String, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let uri_options = Document::deserialize(deserializer)?;
+
+    let mut default_uri_parts = DEFAULT_URI.split('?');
+
+    let mut uri = String::from(default_uri_parts.next().unwrap());
+    // A connection string has two slashes before the host list and one slash before the auth db
+    // name. If an auth db name is not provided the latter slash might not be present, so it needs
+    // to be added manually.
+    if uri.chars().filter(|c| *c == '/').count() < 3 {
+        uri.push('/');
+    }
+    uri.push('?');
+
+    if let Some(options) = default_uri_parts.next() {
+        let options = options.split('&');
+        for option in options {
+            let key = option.split('=').next().unwrap();
+            // The provided URI options should override any existing options in the connection
+            // string.
+            if !uri_options.contains_key(key) {
+                uri.push_str(option);
+                uri.push('&');
+            }
+        }
+    }
+
+    for (key, value) in uri_options {
+        let value = value.to_string();
+        // to_string() wraps quotations around Bson strings
+        let value = value.trim_start_matches('\"').trim_end_matches('\"');
+        uri.push_str(&format!("{}={}&", &key, value));
+    }
+
+    // remove the trailing '&' from the URI (or '?' if no options are present)
+    uri.pop();
+
+    Ok(uri)
+}
+
+fn deserialize_server_api<'de, D>(
```

Comment:
If possible, it would be preferable to just implement Deserialize directly on `ServerApi` to avoid needing to write custom code. From looking at the code here, it looks like we could do it with just an attribute on the `version` field like this (omitting the existing attributes on the struct for brevity):

```rust
#[derive(Deserialize)]
struct ServerApi {
    #[serde(deserialize_with="FromStr::from_str")]
    version: Option<ServerApiVersion>,

    strict: Option<bool>,

    deprecationErrors: Option<bool>, 
}
```

Reply:
This yields:
```
error[E0308]: mismatched types
   --> src/client/options/mod.rs:223:24
    |
223 | #[derive(Clone, Debug, Deserialize, PartialEq, TypedBuilder)]
    |                        ^^^^^^^^^^^ expected `&str`, found type parameter `__D`
    |
    = note:   expected reference `&str`
            found type parameter `__D`
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
   --> src/client/options/mod.rs:223:24
    |
223 | #[derive(Clone, Debug, Deserialize, PartialEq, TypedBuilder)]
    |                        ^^^^^^^^^^^ expected associated type, found struct `error::Error`
    |
    = note: expected associated type `<__D as selection_criteria::_::_serde::Deserializer<'de>>::Error`
                        found struct `error::Error`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider constraining the associated type `<__D as selection_criteria::_::_serde::Deserializer<'de>>::Error` to `error::Error`
    |
223 | #[derive(Clone, Debug, Deserialize<Error = error::Error>, PartialEq, TypedBuilder)]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
   --> src/client/options/mod.rs:223:24
    |
223 | #[derive(Clone, Debug, Deserialize, PartialEq, TypedBuilder)]
    |                        ^^^^^^^^^^^ expected associated type, found struct `error::Error`
    |
    = note: expected associated type `<__D as selection_criteria::_::_serde::Deserializer<'de>>::Error`
                        found struct `error::Error`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider constraining the associated type `<__D as selection_criteria::_::_serde::Deserializer<'de>>::Error` to `error::Error`
    |
223 | #[derive(Clone, Debug, Deserialize<Error = error::Error>, PartialEq, TypedBuilder)]
    |                                   ^^^^^^^^^^^^^^^^^^^^^^
```

Reply:
Ah, sorry, my mistake; I misread what you're doing here. The most idiomatic way to do this would be to implement Deserialize directly on `ServerApiVersion`, which then would allow `Deserialize` to be derived on `ServerApi` without needing any custom code. Here's the code for that (and I promise I made sure it compiles this time!):

```rust
use std::str::FromStr;

use serde::{Deserialize, de::{self, Deserializer, Unexpected}};

pub enum ServerApiVersion {
    Version1,
}

impl<'de> Deserialize<'de> for ServerApiVersion {
    fn deserialize<D>(
        deserializer: D
    ) -> Result<Self, D::Error> where D: Deserializer<'de> {
        let s = String::deserialize(deserializer)?;
        
        ServerApiVersion::from_str(&s)
            .map_err(|_| {
                de::Error::invalid_value(Unexpected::Str(&s), &"a valid version number")
            })
    }
}

#[derive(Deserialize)]
struct ServerApi {
    version: Option<ServerApiVersion>,

    strict: Option<bool>,

    deprecationErrors: Option<bool>, 
}
```

Reply:
Ah, this works now. There was prior art for this (and lots of it at that), but I wasn't sure how to go about this. Thanks for helping out!